### PR TITLE
ch4/xpmem: convert to ch4 am protocols

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Convert atomics from OpenPA to MPL atomics.

This is a split from #4377 since that PR contains too many things.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
The last commit -- "replace chunk counter with offset" was trying to carry out the `TODO` comment:  
```
/* TODO: implement OPA_fetch_and_incr_int uint64_t to directly change cur_offset
        MPI_Aint copy_sz;
     * instead of computing it in each copy */
```
It is included here for discussion -- before I put more effort to fix the bugs within.

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/warnings/quick
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 1]
```diff
@@ -45,56 +45,53 @@
...
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
 
-    total_chunk =
-        data_sz / MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE +
-        (data_sz % MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE == 0 ? 0 : 1);
+    MPIR_Assert(data_sz > 0);
 
-    MPIR_Assert(total_chunk > 1);
+    MPI_Aint cur_offset;
+    MPI_Aint copy_totoal = 0;
```
Comment (KaimingOuyang) [inline src/mpid/ch4/shm/xpmem/xpmem_recv.h]: spell issue for `copy_total`

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 2]
```diff
@@ -45,56 +45,53 @@
...
-    int cur_chunk = 0, total_chunk = 0, num_local_copy = 0;
-    uint64_t cur_offset;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
 
-    total_chunk =
-        data_sz / MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE +
-        (data_sz % MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE == 0 ? 0 : 1);
+    MPIR_Assert(data_sz > 0);
```
Comment (KaimingOuyang) [inline src/mpid/ch4/shm/xpmem/xpmem_recv.h]: let's delete this assert

Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_recv.h]: done.

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_recv.h]: fixed.

**Comment by hzhou** [comment]
test:jenkins/warnings/quick
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/warnings/quick
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

**Comment by hzhou** [comment]
test:jenkins/ch4/xpmem

File: src/mpid/ch4/shm/xpmem/xpmem_impl.h [hunk 3]
```diff
@@ -8,5 +8,199 @@
...
+        MPIR_Comm *comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(sreq, context_id));
+        mpi_errno = MPIDI_XPMEM_send_send_fin(rank, comm, rreq_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+    } else if (fin_type == MPIDI_XPMEM_RECVER_FIN) {
+        /* tell recver we finished so it can free the counter_ptr */
+        int rank = MPIDIG_REQUEST(sreq, rank);
+        MPIR_Comm *comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(sreq, context_id));
+        mpi_errno = MPIDI_XPMEM_send_cnt_free(rank, comm,
+                                              coop_counter_direct_flag, coop_counter_offset);
+        MPIR_ERR_CHECK(mpi_errno);
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_impl.h]: `MPIDI_XPMEM_send_cnt_free` is used in one place. I do not think we need subroutine for it.

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 4]
```diff
@@ -193,40 +119,13 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_coop_recv(uint64_t src_offse
                 " copy dst %p, src %lx, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
                 MPIDIG_REQUEST(rreq, tag), MPIDIG_REQUEST(rreq, context_id),
                 (char *) MPIDIG_REQUEST(rreq, buffer), src_offset, recv_data_sz);
+
     /* sender and receiver datatypes are both continuous, perform cooperative copy. */
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_recv.h]: Empty line here.

File: src/mpid/ch4/shm/xpmem/xpmem_win.c [hunk 5]
```diff
@@ -4,7 +4,7 @@
  *      See COPYRIGHT in top-level directory.
  */
 
-#include "xpmem_impl.h"
+#include "mpidimpl.h"
 #include "xpmem_seg.h"
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_win.c]: Is this change related to coop copy?

File: src/mpid/ch4/shm/xpmem/xpmem_send.h [hunk 6]
```diff
@@ -43,24 +41,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_lmt_isend(const void *buf, MPI_Aint cou
...
-    slmt_req_hdr->tag = tag;
-    slmt_req_hdr->context_id = comm->context_id + context_offset;
-
-    XPMEM_TRACE("lmt_isend: shm ctrl_id %d, src_offset 0x%lx, data_sz 0x%lx, sreq_ptr 0x%lx, "
-                "src_lrank %d, match info[dest %d, src_rank %d, tag %d, context_id 0x%x]\n",
-                MPIDI_SHM_XPMEM_SEND_LMT_RTS, slmt_req_hdr->src_offset,
-                slmt_req_hdr->data_sz, slmt_req_hdr->sreq_ptr, slmt_req_hdr->src_lrank,
-                rank, slmt_req_hdr->src_rank, slmt_req_hdr->tag, slmt_req_hdr->context_id);
-
-    mpi_errno = MPIDI_SHM_do_ctrl_send(rank, comm, MPIDI_SHM_XPMEM_SEND_LMT_RTS, &ctrl_hdr);
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_send.h]: Is the trace message lost?

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 7]
```diff
@@ -68,40 +66,28 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_coop_recv(uint64_t src_offse
...
     }
 
-    XPMEM_TRACE("handle_lmt_coop_recv: shm ctrl_id %d, buf_offset 0x%lx, data_sz 0x%lx, "
-                "sreq_ptr 0x%lx, rreq_ptr 0x%lx, coop_counter_offset 0x%lx, "
-                " local_rank %d, dest %d\n",
-                MPIDI_SHM_XPMEM_SEND_LMT_CTS, slmt_cts_hdr->dest_offset, slmt_cts_hdr->data_sz,
-                slmt_cts_hdr->sreq_ptr, slmt_cts_hdr->rreq_ptr, slmt_cts_hdr->coop_counter_offset,
-                slmt_cts_hdr->dest_lrank, MPIDIG_REQUEST(rreq, rank));
-
     /* Receiver sends CTS packet to sender */
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_recv.h]: Is the trace message lost? Can you please revert it?

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 8]
```diff
@@ -68,40 +66,28 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_coop_recv(uint64_t src_offse
...
         rreq->status.MPI_ERROR = MPI_ERR_TRUNCATE;
 
-    /* XPMEM internal info */
-    slmt_cts_hdr->dest_offset = (uint64_t) MPIDIG_REQUEST(rreq, buffer) + dt_true_lb;
-    slmt_cts_hdr->data_sz = recv_data_sz;
-    slmt_cts_hdr->dest_lrank = MPIDI_XPMEM_global.local_rank;
+    /* allocate shm counter */
+    MPIDI_XPMEM_cnt_t *counter_ptr;
+    counter_ptr = (MPIDI_XPMEM_cnt_t *) MPIR_Handle_obj_alloc(&MPIDI_XPMEM_cnt_mem);
+    MPIDI_XPMEM_REQUEST(rreq, counter_ptr) = counter_ptr;
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_recv.h]: It is very hard to figure out what is changed or moved to a different routine. Can you please avoid unnecessary change such as 
```
MPIDI_XPMEM_REQUEST(rreq, counter_ptr) =
        (MPIDI_XPMEM_cnt_t *) MPIR_Handle_obj_alloc(&MPIDI_XPMEM_cnt_mem);
```
->
```
    MPIDI_XPMEM_cnt_t *counter_ptr;
    counter_ptr = (MPIDI_XPMEM_cnt_t *) MPIR_Handle_obj_alloc(&MPIDI_XPMEM_cnt_mem);
    MPIDI_XPMEM_REQUEST(rreq, counter_ptr) = counter_ptr;
```

File: src/mpid/ch4/shm/xpmem/xpmem_control.h [hunk 9]
```diff
@@ -9,10 +9,15 @@
...
+int MPIDI_XPMEM_rts_cb(int handler_id, void *am_hdr, void **data, size_t * data_sz, int is_local,
+                       int *is_contig, MPIR_Request ** req);
+int MPIDI_XPMEM_cts_cb(int handler_id, void *am_hdr, void **data, size_t * data_sz, int is_local,
+                       int *is_contig, MPIR_Request ** req);
+int MPIDI_XPMEM_send_fin_cb(int handler_id, void *am_hdr, void **data, size_t * data_sz,
+                            int is_local, int *is_contig, MPIR_Request ** req);
+int MPIDI_XPMEM_recv_fin_cb(int handler_id, void *am_hdr, void **data, size_t * data_sz,
+                            int is_local, int *is_contig, MPIR_Request ** req);
+int MPIDI_XPMEM_cnt_free_cb(int handler_id, void *am_hdr, void **data, size_t * data_sz,
+                            int is_local, int *is_contig, MPIR_Request ** req);
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_control.h]: I'd prefer the original function names. In future we may add callback for non-LMT messages, or receiver-initialized LMT. So we need the `send_lmt` prefix.

File: src/mpid/ch4/shm/xpmem/xpmem_types.h [hunk 10]
```diff
@@ -41,23 +35,13 @@ typedef struct MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin {
...
+    uint64_t req_ptr;
+} MPIDI_SHM_ctrl_xpmem_send_lmt_recv_fin_t;
+
 typedef struct MPIDI_SHM_ctrl_xpmem_send_lmt_cnt_free {
     int coop_counter_direct_flag;
     uint64_t coop_counter_offset;
 } MPIDI_SHM_ctrl_xpmem_send_lmt_cnt_free_t;
 
-typedef MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin_t MPIDI_SHM_ctrl_xpmem_send_lmt_recv_fin_t;
-#endif
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_types.h]: I do not see the reason why we need change this line:
```typedef MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin_t MPIDI_SHM_ctrl_xpmem_send_lmt_recv_fin_t;```

File: src/mpid/ch4/shm/xpmem/xpmem_impl.h [hunk 11]
```diff
@@ -20,22 +20,20 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_send_rts(int rank, MPIR_Comm * comm, in
                                                   void *buf, MPI_Aint data_sz,
                                                   int tag, MPIR_Request * sreq)
 {
-    MPIDI_SHM_ctrl_hdr_t ctrl_hdr;
-    MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t *slmt_req_hdr = &ctrl_hdr.xpmem_slmt_rts;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t am_hdr;
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_impl.h]: Renaming `slmt_req_hdr` to `am_hdr` seems unrelated to this commit

File: src/mpid/ch4/shm/xpmem/globals.c [hunk 12]
```diff
@@ -15,12 +15,10 @@ MPL_dbg_class MPIDI_CH4_SHM_XPMEM_GENERAL;
...
 
 /* Preallocated segment objects */
-MPIDI_XPMEM_seg_t MPIDI_XPMEM_seg_mem_direct[MPIDI_XPMEM_SEG_PREALLOC] = { {0}
-};
+MPIDI_XPMEM_seg_t MPIDI_XPMEM_seg_mem_direct[MPIDI_XPMEM_SEG_PREALLOC];
 
 /* Preallocated counter objects */
-MPIDI_XPMEM_cnt_t MPIDI_XPMEM_cnt_mem_direct[MPIDI_XPMEM_CNT_PREALLOC] = { {0}
-};
+MPIDI_XPMEM_cnt_t MPIDI_XPMEM_cnt_mem_direct[MPIDI_XPMEM_CNT_PREALLOC];
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/globals.c]: Is the automatic-zeroed assumption portable cross different platforms? If not, I'd keep the original code for correctness.

File: src/mpid/ch4/shm/xpmem/xpmem_win.c [hunk 13]
```diff
@@ -49,7 +48,7 @@ static int get_node_ranks(MPIR_Comm * shm_comm_ptr, int *shm_ranks, int *node_ra
     goto fn_exit;
 }
 
-int MPIDI_XPMEM_mpi_win_create_hook(MPIR_Win * win)
+int MPIDI_XPMEM_on_win_create(MPIR_Win * win)
 {
```
Comment (minsii) [inline src/mpid/ch4/shm/xpmem/xpmem_win.c]: I'd keep the `_hook` suffix because this is not a full win creation. Same feeling for other routines.

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h [hunk 14]
```diff
@@ -76,17 +76,41 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_lmt_coop_copy(const void *src_buf,
...
+            }
+        } else {
+            /* finish second */
+            if (copy_total == 0) {
+                *fin_type = MPIDI_XPMEM_SENDER_FIN;
+            } else {
+                *fin_type = MPIDI_XPMEM_BOTH_FIN;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_recv.h]: yeah :) I learned that from Kaiming. But hopefully we still can discuss :)

> It hacks MPIDI_XPMEM_do_lmt_coop_copy to distinguish sender and receiver copy

The code passes a flag in argument on whether it is a sender a receiver explicitly, that is not a hack, right?

> It is very hard to fully understanding what are the possible situations and what sync is needed in each situation. The original version gave me a full picture, but it is lost in the revised version.

There are only 4 possible situations listed here. In fact, we could get away with only two situations (which one completes first), but the other two more specific cases are added for optimizations where we can skip some of the synchronizations based on deduction. These are the same four cases the old code has. But the old code use two state variable, a total of 2 x 3 = 6 states, logic split at three places, with two big blocks of comments essentially repeat what the code does, both took me quite a while to understand -- actually, I didn't. I only understood it after I asked and discussed with Kaiming. Do you really think the old code is better?

File: src/mpid/ch4/shm/xpmem/xpmem_control.c [hunk 15]
```diff
@@ -190,38 +189,13 @@ int MPIDI_XPMEM_ctrl_send_lmt_cts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
     }
 
     /* sender and receiver datatypes are both continuous, perform cooperative copy. */
-    mpi_errno =
-        MPIDI_XPMEM_do_lmt_coop_copy(src_buf, slmt_cts_hdr->data_sz, dest_buf,
-                                     &counter_ptr->obj.offset, slmt_cts_hdr->rreq_ptr, sreq,
-                                     TRUE, &fin_type);
+    mpi_errno = MPIDI_XPMEM_do_send(src_buf, slmt_cts_hdr->data_sz, dest_buf,
+                                    &counter_ptr->obj.offset,
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_control.c]: The purpose was to group the code that related to the complicated logic (those four sync states) together. The logic is complicated enough, and the sender-side and recver side need be looked at together to comprehend the logic. So I felt necessity to move the code together.

File: src/mpid/ch4/shm/xpmem/xpmem_impl.h (hunk 3 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_impl.h]: The purpose here is to group all the active message protocol code together. It is easier to read and more importantly easier to maintain. For example, if we decide to switch from internal control protocol to use just generic ch4-layer active message, we only need modify code there. Also in general, shorter code is easier to read, right?

File: src/mpid/ch4/shm/xpmem/xpmem_win.c (hunk 5 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_win.c]: I need check which commit this one belong to. No, the PR moved on and added more refactoring stuff. But please don't be alarmed and there is no pressure to merge any of these change if there is missing agreement. I have been wanting to see xpmem as a simple am protocol rather than as a "pseudo"-shmmod for a while, and this is my trying it out and soliciting discussions. If it works, ofi/ucx all can use the similar model adding lmt support or different variations of lmt support with relative ease.

File: src/mpid/ch4/shm/xpmem/xpmem_send.h (hunk 6 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_send.h]: It does. :) I could add it back, or as I started in another PR, removed all traces. I find these traces are only used during development or debugging and work like scaffold. They can be easily added back on-demand if a debugging session need them and should not be left in production code.

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h (hunk 7 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_recv.h]: I mean, we don't do these traces consistently, e.g. missing in other am protocols, and they are handful to read. Certainly I can add them back if anyone insist.

File: src/mpid/ch4/shm/xpmem/xpmem_recv.h (hunk 8 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_recv.h]: sure.

File: src/mpid/ch4/shm/xpmem/xpmem_control.h (hunk 9 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_control.h]: I was wondering that may be the plan. 

Can we wait until the new components are on the table, then we can look at the actual objects and decide a good naming scheme. Otherwise, we are having these long names (which obviously makes formatting and reading more difficult) for no benefit, right?

File: src/mpid/ch4/shm/xpmem/xpmem_types.h (hunk 10 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_types.h]: Wearing the hat of active message, `send_fin_t` and `recv_fin_t` are two semantically unrelated packet. Aliasing them together not only forces you to use the non-specific member name `req_ptr` rather than `sreq_ptr` and `rreq_ptr` like everywhere else, it also makes two protocols less flexible. For example, if we need pass different amount of information to send_fin_t and recv_fin_t, the aliasing makes it confusing.

In another word, because they are semantically unrelated, I would like the code reflect that semantics.

File: src/mpid/ch4/shm/xpmem/xpmem_impl.h (hunk 11 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_impl.h]: All these routine to send a am packet are similar, declare a local am_hdr structure, then filling the field, and then send them using lower-layer transport. I was thinking that the more we make these code similar, the easier we are able to read them and focus on the details, such as the structure fields and semantics. 

Yes, I can split them or remove it if you prefer.

File: src/mpid/ch4/shm/xpmem/globals.c (hunk 12 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/globals.c]: Static global variables are always initialized to zero, that is in the C standards. I looked it up once :)

We have one situation that compiler will place the global variables to different section and it will cause issue linking with fortran common symbols. In this case, I believe it is safe.

File: src/mpid/ch4/shm/xpmem/xpmem_win.c (hunk 13 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_win.c]: I thought the `_hook` naming is to conform to the netmod/shmmod api interface, and in this case, since xpmem is not really a full shmmod, I felt it helps to differentiate them. Part of the effort in this PR is make xpmem less like a shmmod, but more like an independent utility or protocol module so any shmmod can use.

**Comment by minsii** [comment]
@hzhou Thanks Hui. Of course we can discuss the refactoring code. I do appreciate it and like some of the changes. Since it is not critical, I can take a look at the concerning places again next week (I have to work on several things due end of this week..) and we can continue discussion. Does it work for you?

**Comment by hzhou** [comment]
> Does it work for you?

Yes!

File: src/mpid/ch4/shm/src/shm_control.c [hunk 16]
```diff
@@ -1,41 +0,0 @@
...
-#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
-#include "../xpmem/xpmem_control.h"
-#endif
-
-int MPIDI_SHM_ctrl_dispatch(int ctrl_id, void *ctrl_hdr)
-{
-    int mpi_errno = MPI_SUCCESS;
-
-    switch (ctrl_id) {
-#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
```
Comment (wesbland) [inline src/mpid/ch4/shm/src/shm_control.c]: Removing these functions will break #4331. We need to figure out what to do with that PR before merging this one I think.

Comment (hzhou) [reply src/mpid/ch4/shm/src/shm_control.c]: No problem. Both PRs probably need be looked together at some point.

File: src/mpid/ch4/shm/xpmem/xpmem_control.h (hunk 9 above)
Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_control.h]: Okay, we will add you to the reviewer. The PR won't merge until all the approvals :) You'll have time to get ready

Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/xpmem_control.h]: > We have our own shmmods and following through this format. 

Just for my understanding, you guys have shmmods similar to xpmem that are not really standalone shmmods, but components implements posix shmmod message protocols such as large message transfer, right?

Comment (thananon) [reply src/mpid/ch4/shm/xpmem/xpmem_control.h]: Yes, we are using SHM am ctrl message. We take the design from upstream xpmem mod and make use of the infrastructure that's there. I'm trying to have our code structure consistent with other shmmod (xpmem). 

But since we have a separated mod, the merge is not gonna be difficult. It's just that we will lost this consistency. Which is fine. It's just a bummer. Note that we are targeting 3.4a2, so this won't affect us until the next merge.

I don't have anything against this PR. I think the new name actually make more sense. I just want to voice my concern about the consistency, because we are at the end of the stream. Just a minor nitpick . XD

File: src/mpid/ch4/shm/xpmem/globals.c (hunk 12 above)
Comment (wesbland) [reply src/mpid/ch4/shm/xpmem/globals.c]: I believe the reason we put this in was because it was causing warnings without it (unless I'm remembering incorrectly). If you compile with `-Wall`, do you see any new warnings here?

Comment (hzhou) [reply src/mpid/ch4/shm/xpmem/globals.c]: Here is the strict build log: https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-xpmem/30/compiler=gnu,jenkins_configure=strict,label=xpmem,netmod=ofi,shmmod=xpmem/consoleFull

File: src/mpid/ch4/shm/src/shm_control.c (hunk 16 above)
Comment (wesbland) [reply src/mpid/ch4/shm/src/shm_control.c:18]: We decided to close #4331 so this comment is irrelevant now.

Comment (hzhou) [reply src/mpid/ch4/shm/src/shm_control.c:18]: Cool. @wesbland if you get a chance, could you review this PR anyway?

Comment (wesbland) [reply src/mpid/ch4/shm/src/shm_control.c:18]: I actually already looked at it and this it's generally fine, but I wanted to see the resolution of @minsii 's comments since she had some opinions that were probably more helpful than mine.

**Comment by pavanbalaji** [comment]
This PR doesn't seem to match what the description states.  Also, there's a refactoring going on in the xpmem code to help with the GPU code integration.  So this PR should probably be closed and potentially reopened (after discussion), if desired.

**Comment by hzhou** [comment]
> 
> 
> This PR doesn't seem to match what the description states. Also, there's a refactoring going on in the xpmem code to help with the GPU code integration. So this PR should probably be closed and potentially reopened (after discussion), if desired.

yeah, I am aware that the gpu work is further sliding along the opposite direction of this PR. I'll close it for now and label it for re-open it once we have that discussion.
